lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,299 @@
+
+from pbench.server.db.database import Database
+from sqlalchemy.exc import SQLAlchemyError
+
+
+class DatasetError(Exception):
+    pass
+
+
+class DatasetSqlError(DatasetError):
+    pass
+
+
+class DatasetNotFound(DatasetError):
+    pass
+
+
+class DatasetTransitionError(DatasetError):
+    pass
+
+
+class DatasetTerminalStateViolation(DatasetTransitionError):
+    pass
+
+
+class DatasetBadStateTransition(DatasetTransitionError):
+    pass
+
+
+class States(enum.Enum):
+    """
+    States: Track the progress of a dataset (tarball) through the various
+    stages and operation of the Pbench server.
+    """
+
+    UPLOADING = ("Uploading", True)
+    UPLOADED = ("Uploaded", False)
+    ARCHIVING = ("Archiving", True)
+    ARCHIVED = ("Archived", False)
+    UNPACKING = ("Unpacking", True)
+    UNPACKED = ("Unpacked", False)
+    INDEXING = ("Indexing", True)
+    INDEXED = ("Indexed", False)
+    REINDEXING = ("Re-indexing", True)
+    EXPIRING = ("Expiring", True)
+    EXPIRED = ("Expired", False)
+
+    def __init__(self, friendly, mutating):
+        """
+        __init__ Extended ENUM constructor in order to add a value
+        to record whether each state is a "busy" state where some
+        component is mutating the dataset in some way, and a mixed
+        case "friendly" name for the state. "Mutating" states are
+        expected to be transient as the component should complete
+        the mutation, and should usually have "-ing" endings.
+
+        Args:
+            name (string): Friendly name for the state
+            mutating (boolean): True if a component is mutating dataset
+        """
+        self.friendly = friendly
+        self.mutating = mutating
+
+    def __str__(self):
+        "Return the state's friendly name"
+        return self.friendly
+
+
+class Dataset(Database.Base):
+    """ Identify a Pbench dataset (tarball plus metadata)
+
+    Columns:
+        id          Generated unique ID of table row
+        user        User associated with the dataset
+        controller  Name of controller node
+        name        Base name of dataset (tarball)
+        md5         The dataset MD5 hash (Elasticsearch ID)
+        created     Date the dataset was PUT to server
+        state       The current state of the dataset
+        transition  The timestamp of the last state transition
+
+    TODO: This could be expanded/generalized to support arbitrary user
+    metadata such as search keywords. We need to define some basic
+    parameters and figure out what goes here and what would be more
+    useful in the Elasticsearch DB.
+    """
+
+    __tablename__ = "dataset"
+
+    transitions = {
+        States.UPLOADING: [States.UPLOADED],
+        States.UPLOADED: [States.ARCHIVING],
+        States.ARCHIVING: [States.ARCHIVED],
+        States.ARCHIVED: [States.UNPACKING],
+        States.UNPACKING: [States.UNPACKED],
+        States.UNPACKED: [States.INDEXING],
```

Comment:
Yes, I didn't really try to think about the satellite workflow. It's really going to become a "front end / cache" for the internal server; __but__ that thought raises a really intriguing question we'll have to resolve. Do we replicate the user database between the master and its satellites? Do they have different user databases? If so... how do we figure out what user to use when we pull a dataset from the satellite to the master?

I think I'm officially _not_ thinking about satellites in the context of the PR. They're effectively a `pbench-move-results` which doesn't fit at all in the new world. Your suggestion to change the shim to do a local PUT doesn't really help, either, because that doesn't address the user ownership issue.

Reply:
Let's track this in https://github.com/distributed-system-analysis/pbench/issues/2097.